### ISSUE TITLE
Issues with mad!

### ISSUE THREAD

**Issue by Nosferican** [JuliaStats/StatsBase.jl#366]
I took a look at the code for `mad!` since it is the only piece of code that requires `SpecialFunctions` which introduces `BinDeps` as a dependency. The code used from SpecialFunctions is one line
```
1 / (-sqrt(2 * one(T)) * erfcinv(3 * one(T) / 2))
```
where `T <: Real`. However, `erfcinv` only supports `erfcinv(y::Float64)`, `erfcinv(y::Float32)`, and `erfcinv(x::Integer) = erfcinv(float(x))`. Using `T <: Real` can result in error if `T <: Union{Rational, Irrational, BigFloat, Float16}`. This suggest that we can just hardcode the value as `1.4826022185056018::Float64` and drop those two dependencies.

In regards to the `mad!` function, accepting `T<:Integer` also has issues due to the in-place nature,
```
mad!([1,2,3,4,5,6,8,9])
```
will result in `InexactError()`.

Possible solutions,

1. Strict `!` replaces
 
```
mad!(v::AbstractArray{T};
     center::Real=median!(v),
     normalize::Union{Bool,Nothing}=true,
     constant=nothing) where T<:Union{Float64,Float32}`
```
and make the adjustment either
```
     1.4826022185056018 * median!(v)
```
or
```
    convert(T, 1.4826022185056018) * median!(v)
```

2. May `!` replace

```
mad!(v::AbstractArray{T};
     center::Real=median!(v),
     normalize::Union{Bool,Nothing}=true,
     constant=nothing) where T<:Real`
```
and use `float.(v)` for a special dispatch on `T<:Integer`. Same considerations for hardcoded value apply, but using the updated type.

**Comment by nalimilan** [comment]
Good catch. I think the simplest solution would be to call ` m = median!(v)` first, and then return `oftype(m, 1.4826022185056018) * m`. If we want to be extra careful, we could check that `precision(m)` isn't higher than `precision(Float64)`, in which case the result wouldn't be as precise as it should and better throw an error (what happens currently anyway).

We could also hardcode more digits so that it also works with `BigFloat` (at least with the default precision), but that's strictly an improvement compared to the current code.

Regarding the in-place failure with `Integer`, it's not completely absurd given that there's no way to compute that in-place. I'm not sure whether it's a good idea to allocate a new vector when needed. On the one hand it would allow writing generic code which is efficient when possible, but on the other it would make a potentially large allocation silently.

**Comment by andreasnoack** [comment]
Regarding the first part then I think that hard-coding the constant is actually an opportunity to support `BigFloats`. With the default precision, the `BigFloat` value is `1.482602218505601860547076529360423431326703202590312896536266275245674447622701`.

Regarding the second comment then the `InexactError` is just a risk using `!` versions. The is part of the reason we use `!`.

**Comment by Nosferican** [comment]
I drafted a PR based on the discussion here, but it seems to be running into a segmentation error. Could someone take a look at it?

**Comment by andreasnoack** [comment]
Fixed by #367
